### ISSUE TITLE
Using AndroidAarLibrary will fail on aar containing impl_internal* jar

### ISSUE THREAD

**Issue by mathieubourgeois** [xamarin/xamarin-android#2160]
### Steps to Reproduce

1. In a Xamarin.Android project (not a bindings project) add an AndroidAarLibrary reference to any library containing an impl_internal* internal jar (I had the issue with support-v4-24.1.1.aar from the Android Support library)
2. Try to build the project

### Expected Behavior

Application compiles as usual, aar resources/java code/manifest info are used in the final application.

### Actual Behavior

Compilation fails with the following error :

```
Error: <...>\Xamarin\Android\Xamarin.Android.Common.targets(1196,2): error MSB4018: Échec inattendu de la tâche "ResolveLibraryProjectImports". [<...>]
Error: <...>\Xamarin\Android\Xamarin.Android.Common.targets(1196,2): error MSB4018: System.IO.FileNotFoundException: Le fichier '<...>\internal_impl-24.1.1.jar' est introuvable. [<...>]
Error: <...>\Xamarin\Android\Xamarin.Android.Common.targets(1196,2): error MSB4018: Nom de fichier : '<...>\internal_impl-24.1.1.jar' [<...>]
Error: <...>\Xamarin\Android\Xamarin.Android.Common.targets(1196,2): error MSB4018:    à System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath) [<...>]
Error: <...>\Xamarin\Android\Xamarin.Android.Common.targets(1196,2): error MSB4018:    à System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost) [<...>]
Error: <...>\Xamarin\Android\Xamarin.Android.Common.targets(1196,2): error MSB4018:    à System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access) [<...>]
Error: <...>\Xamarin\Android\Xamarin.Android.Common.targets(1196,2): error MSB4018:    à Xamarin.Android.Tools.Files.HashFile(String filename, HashAlgorithm hashAlg) [<...>]
Error: <...>\Xamarin\Android\Xamarin.Android.Common.targets(1196,2): error MSB4018:    à Xamarin.Android.Tools.Files.HashString(String s) [<...>]
Error: <...>\Xamarin\Android\Xamarin.Android.Common.targets(1196,2): error MSB4018:    à Xamarin.Android.Tasks.ResolveLibraryProjectImports.<>c.<Extract>b__60_9(String entryFullName) [<...>]
Error: <...>\Xamarin\Android\Xamarin.Android.Common.targets(1196,2): error MSB4018:    à Xamarin.Android.Tools.Files.ExtractAll(ZipArchive zip, String destination, Action`2 progressCallback, Func`2 modifyCallback, Func`2 deleteCallback, Boolean forceUpdate) [<...>]
Error: <...>\Xamarin\Android\Xamarin.Android.Common.targets(1196,2): error MSB4018:    à Xamarin.Android.Tasks.ResolveLibraryProjectImports.Extract(DirectoryAssemblyResolver res, ICollection`1 jars, ICollection`1 resolvedResourceDirectories, ICollection`1 resolvedAssetDirectories, ICollection`1 resolvedEnvironments) [<...>]
Error: <...>\Xamarin\Android\Xamarin.Android.Common.targets(1196,2): error MSB4018:    à Xamarin.Android.Tasks.ResolveLibraryProjectImports.Execute() [<...>]
Error: <...>\Xamarin\Android\Xamarin.Android.Common.targets(1196,2): error MSB4018:    à Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [<...>]
Error: <...>\Xamarin\Android\Xamarin.Android.Common.targets(1196,2): error MSB4018:    à Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext() [<...>]
```

### Extra informations

- The same project using `LibraryProjectZip` in a fake bindings project works fine (but we don't need the bindings, so an extra dll/so that can be avoided is a good thing)
- I took a look at the code that seems to be causing the issue, it comes from the initial implementation of `AndroidAarLibrary` in `ResolveLibraryProjectImport.cs` :

```
				using (var zip = MonoAndroidHelper.ReadZipFile (aarFile.ItemSpec)) {
					try {
						updated |= Files.ExtractAll (zip, importsDir, modifyCallback: (entryFullName) => {
							var entryFileName = Path.GetFileName (entryFullName);
							var entryPath = Path.GetDirectoryName (entryFullName);
							if (entryFileName.StartsWith ("internal_impl", StringComparison.InvariantCulture)) {
								var hash = Files.HashString (entryFileName);
								return Path.Combine (entryPath, $"internal_impl-{hash}.jar");
							}
							return entryFullName;
						}, deleteCallback: (fileToDelete) => {
							return !jars.Contains (fileToDelete);
						}, forceUpdate: false);

						if (Directory.Exists (importsDir) && (updated || !stamp.Exists)) {
							Log.LogDebugMessage ("Touch {0}", stamp.FullName);
							stamp.Create ().Close ();
						}
					} catch (PathTooLongException ex) {
						Log.LogErrorFromException (new PathTooLongException ($"Error extracting resources from \"{aarFile.ItemSpec}\"", ex));
					}
				}
``` 

From what I can see, it extracts the aar file and needs to check the hashes of the internal jar to see if something actually changed (my initial guess at least). However, the commit (6a8ea2bbdbad65f1f46ba60a343506c92d1e0a1e) doesn't mention anything about this so I'm not entirely sure if that's supposed to be the case.

So, when it finds internal_impl* in the aar, it tries to hash `"internal_impl*"` file without any path. However, this means it checks at the base of the project (or the cwd I would guess) for that file, but there's no reason for it to be there, I don't think it would be extracted there anyway) Therefore, we have the crash.

I guess the fix is probably to check two things
a) consider the proper path to the unzipped file
b) consider an empty hash if the unzipped file doesn't exist yet

### Version Information

Microsoft Visual Studio Professional 2017 
Version 15.8.1
VisualStudio.15.Release/15.8.1+28010.2003
Microsoft .NET Framework
Version 4.7.03062

Installed Version: Professional

Visual C++ 2017   00370-20007-36350-AA982
Microsoft Visual C++ 2017

Application Insights Tools for Visual Studio Package   8.13.10627.1
Application Insights Tools for Visual Studio

ASP.NET and Web Tools 2017   15.8.05074.0
ASP.NET and Web Tools 2017

ASP.NET Core Razor Language Services   15.8.31590
Provides languages services for ASP.NET Core Razor.

ASP.NET Web Frameworks and Tools 2017   5.2.60618.0
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   15.8.05023.0
Azure App Service Tools v3.0.0

C# Tools   2.9.0-beta8-63208-01
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Cookiecutter   15.8.18201.1
Provides tools for finding, instantiating and customizing templates in cookiecutter format.

JavaScript Language Service   2.0
JavaScript Language Service

JavaScript Project System   2.0
JavaScript Project System

Merq   1.1.38 (5b3c069)
Command Bus, Event Stream and Async Manager for Visual Studio extensions.

Microsoft Azure Tools   2.9
Microsoft Azure Tools for Microsoft Visual Studio 2017 - v2.9.10730.2

Microsoft Continuous Delivery Tools for Visual Studio   0.4
Simplifying the configuration of continuous build integration and continuous build delivery from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft Library Manager   1.0
Install client-side libraries easily to any web project

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual C++ Wizards   1.0
Microsoft Visual C++ Wizards

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

Microsoft Visual Studio VC Package   1.0
Microsoft Visual Studio VC Package

MLGen Package Extension   1.0
MLGen Package Visual Studio Extension Detailed Info

Mono Debugging for Visual Studio   4.11.7-pre (8955b2a)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   4.6.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit http://docs.nuget.org/.

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

Python   15.8.18201.1
Provides IntelliSense, projects, templates, debugging, interactive windows, and other support for Python developers.

Python - Django support   15.8.18201.1
Provides templates and integration for the Django web framework.

Python - IronPython support   15.8.18201.1
Provides templates and integration for IronPython-based projects.

Python - Profiling support   15.8.18201.1
Profiling support for Python projects.

ResourcePackage Extension   1.0
ResourcePackage Visual Studio Extension Detailed Info

ResourcePackage Extension   1.0
ResourcePackage Visual Studio Extension Detailed Info

SQL Server Data Tools   15.1.61808.07020
Microsoft SQL Server Data Tools

Test Adapter for Boost.Test   1.0
Enables Visual Studio's testing tools with unit tests written for Boost.Test.  The use terms and Third Party Notices are available in the extension installation directory.

Test Adapter for Google Test   1.0
Enables Visual Studio's testing tools with unit tests written for Google Test.  The use terms and Third Party Notices are available in the extension installation directory.

TypeScript Tools   15.8.20801.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   2.9.0-beta8-63208-01
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual C++ for Cross Platform Mobile Development (Android)   15.0.27924.00
Visual C++ for Cross Platform Mobile Development (Android)

Visual C++ for Cross Platform Mobile Development (iOS)   15.0.27825.00
Visual C++ for Cross Platform Mobile Development (iOS)

Visual F# Tools 10.2 for F# 4.5   15.8.0.0.  Commit Hash: c55dd2c3d618eb93a8d16e503947342b1fa93556.
Microsoft Visual F# Tools 10.2 for F# 4.5

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Tools for CMake   1.0
Visual Studio Tools for CMake

Visual Studio Tools for Containers   1.0
Visual Studio Tools for Containers

Visual Studio Tools for Unity   3.8.0.3
Visual Studio Tools for Unity

Visual Studio Tools for Universal Windows Apps   15.0.28010.00
The Visual Studio Tools for Universal Windows apps allow you to build a single universal app experience that can reach every device running Windows 10: phone, tablet, PC, and more. It includes the Microsoft Windows 10 Software Development Kit.

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

Xamarin   4.11.0.732 (d15-8@33e83e124)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   4.14.218 (79f535bdd)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin Templates   1.1.113 (e1d02a7)
Templates for building iOS, Android, and Windows apps with Xamarin and Xamarin.Forms.

Xamarin.Android SDK   9.0.0.18 (HEAD/3d8a28f1a)
Xamarin.Android Reference Assemblies and MSBuild support.

Xamarin.iOS and Xamarin.Mac SDK   11.14.0.13 (373c313)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.

**Comment by mathieubourgeois** [comment]
Bump?

I would really like to fix this bug, however I don't see any actual use case for internal_impl in xamarin android. I could just remove


if (entryFileName.StartsWith ("internal_impl", StringComparison.InvariantCulture)) {
	var hash = Files.HashString (entryFileName);
	return Path.Combine (entryPath, $"internal_impl-{hash}.jar");
}

because it looks like it could never have worked before. However, I think this was added there for a reason, so I'd prefer to fix it instead of just removing the offending code.  Do we want to hash the file we're about to extract? Why do we actually need to change the name of that jar for it's hash instead of the version (usually, it's internal_impl-{versionName}.jar) Is there any actual use case for that code?

**Comment by mathieubourgeois** [comment]
This seems to have been fixes indiretly by c362fe571dd96170a9e3e7e6134e72b6f957721d, the `HashString` function that is called in that case now only hash the name of the file, instead of the content of the file, which seems to work properly. Therefore closing the issue.
